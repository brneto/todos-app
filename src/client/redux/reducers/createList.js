import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import { prop, identity, equals, always } from 'ramda';
import { produce } from 'immer';
import { createSelector } from 'reselect';
import { commands, events, documents } from '../actions';

const createList = filter => {
  const isFilter = equals(filter);

  const ids = handleActions(
      {
        [documents.todosFetched]: {
          next: produce((draft, { payload, meta }) => {
            if (isFilter(meta.filter)) return payload.result; //return an entirely new state
          }),
        },
        [documents.todoAdded]: {
          next: produce((draft, { payload }) => {
            if (!isFilter('completed')) draft.push(payload.result); //modify the current draft state
          }),
        },
        [commands.addTodoToList]: {
          next: produce((draft, { payload, meta }) => {
            if (isFilter(meta.filter)) draft.push(payload.result); //modify the current draft state
          }),
        },
        [commands.removeTodoFromList]: {
          next: produce((draft, { payload, meta }) => {
            const notEquals = a => b => a !== b;
            if(isFilter(meta.filter)) return draft.filter(notEquals(payload.result)); //return an entirely new state
          }),
        },
      },
      [] // Initial state
    );

  const isFetching = handleActions(
      {
        [events.fetchingTodos]: produce((draft, { payload }) => {
          if (isFilter(payload)) return true; //return an entirely new state
        }),
        [events.fetchedTodos]: produce((draft, { payload }) => {
          if (isFilter(payload)) return false;
        }),
      },
      false // Initial state
    );

  const
    getErrorMessage = (test, { message }) => {
      if (test) return (message ?? 'Something went wrong.');
    },
    errorMessage = handleActions(
      {
        [documents.todosFetched]: {
          next: always(null),
          throw: produce((draft, { payload, meta }) =>
            getErrorMessage(isFilter(meta.filter), payload) //return an entirely new state
          )
        },
        [documents.todoAdded]: {
          next: always(null),
          throw: produce((draft, { payload }) =>
            getErrorMessage(!isFilter('completed'), payload) //return an entirely new state
          )
        },
        [documents.todoToggled]: {
          next: always(null),
          throw: produce((state, { payload }) =>
            getErrorMessage(isFilter('all'), payload) //return an entirely new state
          )
        },
      },
      null // Initial state
    );

  return combineReducers({
    ids,
    isFetching,
    errorMessage,
  });
};

// SELECTORS
// prop :: s -> {s: a} -> a | Undefined
const
  getIds = createSelector([prop('ids')], identity),
  getIsFetching = createSelector([prop('isFetching')], identity),
  getErrorMessage = createSelector([prop('errorMessage')], identity);

export {
  createList as default,
  getIds,
  getIsFetching,
  getErrorMessage,
};
