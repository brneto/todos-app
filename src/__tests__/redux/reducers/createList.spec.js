import deepFreeze from 'deep-freeze';
import createList from '../../../client/redux/reducers/createList';
import {
  setFetchTodos,
  setAddTodo,
  setToggleTodo,
  setToggleFetching
} from '../../../client/redux/actions';

describe('reducers/createList', () => {
  const filter = 'active';
  const listByFilter = createList(filter);
  const initialState = deepFreeze(listByFilter(undefined, { type: 'INIT' }));
  const baseData = { result: [1, 2, 3] };
  const baseState = deepFreeze(
    listByFilter(initialState, setFetchTodos(baseData, filter))
  );
  const error = new Error('Boom!');

  it('should handle unknown actions', () => {
    expect(listByFilter(initialState, { type: 'FAKE' })).toBe(initialState);
  });

  it('should add fetched todos to empty list', () => {
    expect(
      listByFilter(initialState, setFetchTodos(baseData, filter))
    ).toMatchSnapshot();
  });

  it('should add fetched todos to non-empty list', () => {
    const testData = { result: [4, 5, 6] };

    expect(
      listByFilter(baseState, setFetchTodos(testData, filter))
    ).toMatchSnapshot();
  });

  it('should add fetched todos saves error message', () => {
    expect(listByFilter(baseState, setFetchTodos(error, filter)))
    .toMatchSnapshot();
  });

  it('should add todo to empty list', () => {
    const testData = { result: 4 };

    expect(
      listByFilter(initialState, setAddTodo(testData, filter))
    ).toMatchSnapshot();
  });

  it('should add todo to non-empty list', () => {
    const testData = { result: 4 };

    expect(
      listByFilter(baseState, setAddTodo(testData, filter))
    ).toMatchSnapshot();
  });

  it('should add todo save error message', () => {
    expect(listByFilter(baseState, setAddTodo(error)))
    .toMatchSnapshot();
  });

  it('should toggle todo remove one id from list', () => {
    const testData = {
      entities: {
        todos: {
          1: {
            id: 1,
            text: 'Learn Redux',
            completed: true
          }
        }
      },
      result: 1
    };

    expect(listByFilter(baseState, setToggleTodo(testData, filter)))
    .toMatchSnapshot();
  });

  it('should toggle todo saves error message', () => {
    const filter = 'all';
    const listByFilter = createList(filter);
    const initialState = deepFreeze(listByFilter(undefined, { type: 'INIT' }));
    const baseData = { result: [1, 2, 3] };
    const baseState = deepFreeze(
      listByFilter(initialState, setFetchTodos(baseData, filter))
    );

    expect(listByFilter(baseState, setToggleTodo(error)))
    .toMatchSnapshot();
  });

  it('should toggle fetching parameter', () => {
    const firstState = listByFilter(baseState, setToggleFetching(filter));
    expect(firstState.isFetching).toBe(true);

    const secondState = listByFilter(firstState, setToggleFetching(filter));
    expect(secondState.isFetching).toBe(false);
  });
});
