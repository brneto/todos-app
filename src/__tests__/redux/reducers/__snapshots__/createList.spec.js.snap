// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/createList should add a todo to the list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    1,
    2,
    3,
    5,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add fetched todos saves error message 1`] = `
Object {
  "errorMessage": "Boom!",
  "ids": Array [
    1,
    2,
    3,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add fetched todos to empty list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    1,
    2,
    3,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add fetched todos to non-empty list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    4,
    5,
    6,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add todo save error message 1`] = `
Object {
  "errorMessage": "Boom!",
  "ids": Array [
    1,
    2,
    3,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add todo to empty list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    4,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should add todo to non-empty list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    1,
    2,
    3,
    4,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should remove a todo from the list 1`] = `
Object {
  "errorMessage": null,
  "ids": Array [
    2,
    3,
  ],
  "isFetching": false,
}
`;

exports[`reducers/createList should toggle todo saves error message 1`] = `
Object {
  "errorMessage": "Boom!",
  "ids": Array [
    1,
    2,
    3,
  ],
  "isFetching": false,
}
`;
